mm2src/kdf_walletconnect/src/lib.rs

```diff
@@ -0,0 +1,560 @@
+        let namespaces = match namespaces {
+            Some(value) => Some(serde_json::from_value(value)?),
+            None => None,
+        };
+        let (topic, url) = self.pairing.create(self.metadata.clone(), None)?;
+
+        info!("[topic] Subscribing to topic");
+
+        for attempt in 0..MAX_RETRIES {
+            match self
+                .client
+                .subscribe(topic.clone())
+                .timeout_secs(PUBLISH_TIMEOUT_SECS)
+                .await
+            {
+                Ok(Ok(_)) => {
+                    info!("[topic] Subscribed to topic");
+                    send_proposal_request(self, &topic, namespaces).await?;
+                    return Ok(url);
+                },
+                Ok(Err(err)) => return MmError::err(err.into()),
+                Err(_) => self.wait_until_client_is_online_loop(attempt).await,
+            }
+        }
+
+        MmError::err(WalletConnectError::InternalError(
+            "client connection timeout".to_string(),
+        ))
+    }
+
+    /// Retrieves the symmetric key associated with a given `topic`.
+    fn sym_key(&self, topic: &Topic) -> MmResult<SymKey, WalletConnectError> {
+        if let Some(key) = self.session.sym_key(topic) {
+            return Ok(key);
+        }
+
+        if let Ok(key) = self.pairing.sym_key(topic) {
+            return Ok(key);
+        }
+
+        MmError::err(WalletConnectError::InternalError(format!(
+            "topic sym_key not found:{topic}"
+        )))
+    }
+
+    /// Handles an inbound published message by decrypting, decoding, and processing it.
+    async fn handle_published_message(&self, msg: PublishedMessage) -> MmResult<(), WalletConnectError> {
+        let message = {
+            let key = self.sym_key(&msg.topic)?;
+            decode_and_decrypt_type0(msg.message.as_bytes(), &key)?
+        };
+
+        info!("[{}] Inbound message payload={message}", msg.topic);
+
+        match serde_json::from_str(&message)? {
+            Payload::Request(request) => process_inbound_request(self, request, &msg.topic).await?,
+            Payload::Response(response) => process_inbound_response(self, response, &msg.topic).await,
+        }
+
+        info!("[{}] Inbound message was handled successfully", msg.topic);
+
+        Ok(())
+    }
+
+    /// Loads sessions from storage, activates valid ones, and deletes expired ones.
+    async fn load_session_from_storage(&self) -> MmResult<(), WalletConnectError> {
+        info!("Loading WalletConnect session from storage");
+        let now = chrono::Utc::now().timestamp() as u64;
+        let mut sessions = self
+            .session
+            .storage()
+            .get_all_sessions()
+            .await
+            .mm_err(|err| WalletConnectError::StorageError(err.to_string()))?;
+
+        // bring most recent active session to the back.
+        sessions.sort_by(|a, b| a.expiry.cmp(&b.expiry));
```

Comment:
why? is that to set the most recent session as the active? could do so without sorting.

Reply:
the most recent session is the one with the farthest expiration time

Reply:
we could choose that with maxing instead of sorting, that's cheaper.
is the sorting needed for anything else? i.e. do we need to keep the sessions sorted in some way for any reason?

Reply:
since the data are already sorted in db there's no need to sort, I can just map over them in reversed order.